### DISCUSSION TITLE
RTL-SDR Blog V4 requires newer driver

### DISCUSSION THREAD

**Discussion by bdr99** [hertzg/rtl_433_docker#106] [General]
I'm using an [RTL-SDR Blog V4](https://www.amazon.com/dp/B0CD7558GT) kit and it seems this model requires a newer driver, as described on [this page](https://www.rtl-sdr.com/V4/).

If I attempt to use the driver bundled with rtl_433, it fails with this output:

```
$ docker run --rm --device /dev/bus/usb/002/002 hertzg/rtl_433:23.11-debian

rtl_433 version 23.11 (2023-11-28) inputs file rtl_tcp RTL-SDR SoapySDR with TLS
Detached kernel driver
Found Rafael Micro R828D tuner
Exact sample rate is: 250000.000414 Hz
[SDR] Using device 0: RTLSDRBlog, Blog V4, SN: 00000001, "Generic RTL2832U OEM"
[R82XX] PLL not locked!
[R82XX] PLL not locked!
Allocating 15 zero-copy buffers
```

Following the instructions on the above linked page makes the device work correctly, but I'm not sure how to integrate the newer driver into this Docker image. Any assistance would be appreciated.

**Comment by hertzg** [comment]
You would need to build the image with new drivers, that means you need to adjust the build stage to not install the libsdr with `apk` and follow the guide you provided.

https://github.com/hertzg/rtl_433_docker/blob/aa84c25b03775f2aef4a6c8061113f3a97b28953/images/debian/build-context/Dockerfile#L4-L24

Then modify the runtime to have runtime environment to have those build libraries

https://github.com/hertzg/rtl_433_docker/blob/aa84c25b03775f2aef4a6c8061113f3a97b28953/images/debian/build-context/Dockerfile#L34-L43

If you want to experiment I'd suggest running a test contianer and comming up with the correct set of commands to then build the correct Dockerfile.

```
$ docker run --rm -it --device /dev/bus/usb debian:bookworm
```

this will give you a "playground" container where you can experiment with building the lib and app correctly and also test the compiled executable safely.

**Reply by bdr99** [reply]
Thanks for the advice. I'll give this a try. It would also be great to get this driver added to the image on docker hub for everyone, that way others with the V4 device won't run into the same issues I did. I expect that more people will be buying the V4 device since it's newer and has more features compared to V3.

**Reply by bdr99** [reply]
@hertzg I was able to get it my V4 device to work with these changes to the Debian Dockerfile. Would you accept a PR with these changes?

```diff
diff --git a/images/debian/build-context/Dockerfile b/images/debian/build-context/Dockerfile
index 15662fe..7a6d279 100644
--- a/images/debian/build-context/Dockerfile
+++ b/images/debian/build-context/Dockerfile
@@ -4,13 +4,23 @@ FROM debian:${debianVersion} as builder
 RUN apt-get update && apt-get install -y \
     build-essential \
     cmake \
+    debhelper \
     git \
     libusb-1.0-0-dev \
     libsoapysdr-dev \
     libssl-dev \
-    librtlsdr-dev \
+    pkg-config \
  && rm -rf /var/lib/apt/lists/*

+WORKDIR /rtl-sdr
+RUN git clone https://github.com/rtlsdrblog/rtl-sdr-blog && \
+    cd rtl-sdr-blog && \
+    dpkg-buildpackage -b --no-sign && \
+    cd .. && \
+    rm -rf rtl-sdr-blog && \
+    dpkg -i librtlsdr0_*.deb && \
+    dpkg -i librtlsdr-dev_*.deb
+
 WORKDIR /build

 ADD ./rtl_433 ./rtl_433
@@ -33,12 +43,15 @@ LABEL maintainer="georgedot@gmail.com" \

 RUN apt-get update && apt-get install -y \
     libusb-1.0-0 \
-    librtlsdr0 \
     '^libsoapysdr0\.[6-8]$' \
     '^libssl(1\.1|3)$' \
     '^soapysdr0\.[6-8]-module-all$' \
  && rm -rf /var/lib/apt/lists/*

+COPY --from=builder /rtl-sdr/librtlsdr0_*.deb /tmp/
+RUN dpkg -i /tmp/librtlsdr0_*.deb && \
+    rm -rf /tmp/librtlsdr0_*.deb
+
 WORKDIR /root
 COPY --from=builder /build/root/ /
```

**Reply by hertzg** [reply]
Does this work on non v4 devices?

**Reply by droans** [reply]
Per the [documentation](https://www.rtl-sdr.com/v4/), the new drivers should be backwards compatible. I don't have any other adapter to test with.

I've been running my own custom Docker build for a while now using the drivers and I've had no issues.

**Reply by bdr99** [reply]
@hertzg Sorry for the delayed response. I have both a V3 and V4 device, and I have tested this change and it works with both.

**Comment by hertzg** [comment]
Converting this to a discussion since this does not look like an issue with the image itself.

**Comment by Toxicable** [comment]
I just ran into this one, applied the above patch and now my docker container appears to be working, although I'm not getting any output, but that could be another issue

Before patch:
```
docker run --device /dev/bus/usb/001/008 hertzg/rtl_433 -f 433.92M -v
...
[SDR] Using device 0: RTLSDRBlog, Blog V4, SN: 00000001, "Generic RTL2832U OEM"
Exact sample rate is: 250000.000414 Hz
[SDR] Sample rate set to 250000 S/s.
[Input] Bit detection level set to 0.0 (Auto).
[SDR] Tuner gain set to Auto.
[Input] Reading samples in async mode...
[SDR] rtlsdr_set_center_freq 433920000 = -1
[sdr_set_center_freq] Failed to set center freq.
```
After
```
docker run --device /dev/bus/usb/001/008 sha256:324596453ec01458273e1f153a09e0f287f628196bcc33ddeb41ebf378cabc68 -v -f 433.92M
...
[SDR] Using device 0: RTLSDRBlog, Blog V4, SN: 00000001, "Generic RTL2832U OEM"
Exact sample rate is: 250000.000414 Hz
[SDR] Sample rate set to 250000 S/s.
[Input] Bit detection level set to 0.0 (Auto).
[SDR] Tuner gain set to Auto.
[Input] Reading samples in async mode...
[SDR] Tuned to 433.920MHz.
[Baseband] low pass filter for 250000 Hz at cutoff 25000 Hz, 40.0 us
```

**Reply by droans** [reply]
What's the rtl-433 version shown in the logs? 

I was running into a similar issue. I discovered that the default version of rtl-433 in the Debian repository was 21.12-1, but upgrading to 23.11-1+b1 fixed any issues I had.

**Comment by handly-intel** [comment]
Hi everyone, and thanks for maintaining this project.

I am seeing a very similar `R82XX PLL not locked` behavior, but on a different setup (not RTL-SDR Blog V4), so I wanted to ask if this might still be related to the same driver/runtime family of issues.

### Environment
- Board: Banana Pi BPI-M4-Zero (Allwinner H618)
- OS: Armbian Linux 6.12.58-current-sunxi64
- Arch: aarch64
- Images tested:
  - `hertzg/rtl_433:latest`
  - `hertzg/rtl_433:latest-debian-trixie`
- SDR dongle: Generic RTL2838U (`0bda:2838`)
- Stack: Mosquitto + Node-RED

### What happens
Container starts and detects the tuner, but logs:
- `Found Rafael Micro R820T tuner`
- `Exact sample rate is: 250000.000414 Hz`
- `[R82XX] PLL not locked!`

In this state, the container does not decode data as expected.

### Notes
- USB passthrough is configured (`/dev/bus/usb`) and container is privileged.
- Host detects the dongle normally (`lsusb`).
- I already checked for common conflicts and restart/replug scenarios.
- Frequency context is 433 MHz.
- This dongle is **not** RTL-SDR Blog V4.

Could this be a related `librtlsdr` compatibility issue on ARM64 even for non-V4 hardware?  
If helpful, I can open a separate issue with full logs and exact repro steps.

Thanks again for your time and guidance.
